### PR TITLE
PWGHF: Add Bs tree creator

### PR DESCRIPTION
**Pull request by pstahlhu**
First implementation of Bs tree creator @fcatalan92 @fgrosa.

### CODE REVIEW

**Review by fcatalan92 — COMMENTED** [review]
@pstahlhu Thanks a lot for this PR! It looks fine to me, I have added just a couple of code suggestions.

**Review by fcatalan92 — APPROVED** [review]
@vkucera Ready to be merged from my side

File: PWGHF/TableProducer/treeCreatorBsToDsPi.cxx [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    }
+  }
+
+  void processData(aod::Collisions const& collisions,
+                   soa::Filtered<soa::Join<aod::HfCandBs, aod::HfSelBsToDsPi>> const& candidates,
+                   TracksWPid const&)
+  {
+    // Filling event properties
+    rowCandidateFullEvents.reserve(collisions.size());
+    for (auto const& collision : collisions) {
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorBsToDsPi.cxx:280]:
```suggestion
    for (const auto& collision : collisions) {
```

File: PWGHF/TableProducer/treeCreatorBsToDsPi.cxx [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+      fillEvent(collision, 0, 1);
+    }
+
+    // Filling candidate properties
+    rowCandidateFull.reserve(candidates.size());
+    if (fillCandidateLiteTable) {
+      rowCandidateLite.reserve(candidates.size());
+    }
+    for (auto const& candidate : candidates) {
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorBsToDsPi.cxx:289]:
```suggestion
    for (const auto& candidate : candidates) {
```

File: PWGHF/TableProducer/treeCreatorBsToDsPi.cxx [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+  void processMc(aod::Collisions const& collisions,
+                 aod::McCollisions const&,
+                 SelectedCandidatesMc const& candidates,
+                 soa::Join<aod::McParticles, aod::HfCandBsMcGen> const& particles,
+                 TracksWPid const&)
+  {
+    // Filling event properties
+    rowCandidateFullEvents.reserve(collisions.size());
+    for (auto const& collision : collisions) {
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorBsToDsPi.cxx:311]:
```suggestion
    for (const auto& collision : collisions) {
```

File: PWGHF/TableProducer/treeCreatorBsToDsPi.cxx [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+      }
+      for (const auto& candidate : candidates) {
+        auto prong1 = candidate.prong1_as<TracksWPid>();
+        fillCandidateTable<true>(candidate, prong1);
+      }
+    }
+
+    // Filling particle properties
+    rowCandidateFullParticles.reserve(particles.size());
+    for (auto const& particle : particles) {
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorBsToDsPi.cxx:351]:
```suggestion
    for (const auto& particle : particles) {
```

File: PWGHF/TableProducer/treeCreatorBsToDsPi.cxx [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    if (fillCandidateLiteTable) {
+      rowCandidateLite.reserve(candidates.size());
+    }
+    for (auto const& candidate : candidates) {
+      if (fillOnlyBackground) {
+        float pseudoRndm = candidate.ptProng1() * 1000. - (int64_t)(candidate.ptProng1() * 1000);
+        if (candidate.pt() < ptMaxForDownSample && pseudoRndm >= downSampleBkgFactor) {
+          continue;
+        }
+      }
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorBsToDsPi.cxx:295]:
```suggestion
      if (fillOnlyBackground && downSampleBkgFactor < 1.) {
        float pseudoRndm = candidate.ptProng1() * 1000. - (int64_t)(candidate.ptProng1() * 1000);
        if (pseudoRndm >= downSampleBkgFactor && candidate.pt() < ptMaxForDownSample) {
          continue;
        }
      }
```
This should avoid unnecessary calculations.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by pstahlhu** [comment]
Thank you for your comments @vkucera !

**Review by vkucera — APPROVED** [review]

**Pull request merged**